src/index.ts

```diff
@@ -1,89 +1,148 @@
-const kafkaHost = process.env.KAFKA_HOST || "http://localhost";
-const kafkaPort = process.env.KAFKA_PORT || "9092";
+const kafkaHost = process.env.KAFKA_HOST ?? 'localhost'
+const kafkaPort = process.env.KAFKA_PORT ?? '9092'
+const trustSelfSigned = (process.env.TRUST_SELF_SIGNED ?? 'false') === 'true'
+const openhimAuth = {
+  apiURL: process.env.OPENHIM_API_URL ?? 'https://localhost:8080',
+  username: process.env.OPENHIM_USERNAME ?? 'root@openhim.org',
+  password: process.env.OPENHIM_PASSWORD ?? 'password',
+  trustSelfSigned,
+  urn: 'urn:openhim-mediator:kafka-mapper-consumer'
+}
+const shouldRegisterMediator =
+  (process.env.REGISTER_MEDIATOR ?? 'false') === 'true'
+const fromBeginning = (process.env.KAFKA_FROM_BEGINNING ?? 'true') === 'true'
+const consumerGroupId = process.env.CONSUMER_GROUP_ID ?? 'kafka-mapper-consumer'
 
 const kafka = new Kafka({
   logLevel: logLevel.INFO,
   brokers: [`${kafkaHost}:${kafkaPort}`],
-  clientId: "kafka-mapper-consumer",
-});
+  clientId: 'kafka-mapper-consumer'
+})
+
+let consumer = kafka.consumer({groupId: 'kafka-mapper-consumer'})
+
+const run = async (newFhirMappings?) => {
+  if (newFhirMappings) {
+    await consumer.disconnect()
+    // create new consumer so we can subscribe to a (potentially) different set of topics
+    consumer = kafka.consumer({groupId: consumerGroupId})
+    fhirMappings = newFhirMappings
+  }
 
-const consumer = kafka.consumer({ groupId: "kafka-mapper-consumer" });
-const producer = kafka.producer();
+  // Register as an OpenHIM mediator and use that for config management
+  if (shouldRegisterMediator && !newFhirMappings) {
+    console.log(
+      'Registering as an OpenHIM Mediator. The OpenHIM will be used to supply latest config not the config file, the config file will provide the default config to the OpenHIM.'
+    )
+    await promisify(registerMediator)(openhimAuth, mediatorConfig)
+    const configEmitter = activateHeartbeat(openhimAuth)
 
-const run = async () => {
-  const topics = fhirMappings.map((fhirMapping) => fhirMapping.resourceType.toLowerCase());
+    configEmitter.on('config', config => {
+      console.log('Received new mapping config, restarting consumer...')
+      // restart consumer with new config
+      run(JSON.parse(config.fhirMappings))
+    })
 
-  await consumer.connect();
-  await producer.connect();
-  await consumer.subscribe({ topics, fromBeginning: true });
+    const config = await promisify(fetchConfig)(openhimAuth)
+    // set config to that received from the OpenHIM
+    fhirMappings = JSON.parse(config.fhirMappings)
+  }
+
+  const fhirMappingValidationErrors = ValidateFhirMappingsJson(fhirMappings)
+  if (fhirMappingValidationErrors.length > 0) {
+    console.error('Invalid fhir-mapping.json file')
+    fhirMappingValidationErrors.forEach(error => console.error(error))
+    process.exit(1)
+  }
+
+  const topics = fhirMappings.map(fhirMapping =>
+    fhirMapping.resourceType.toLowerCase()
+  )
+
+  await consumer.connect()
```

Comment:
We may need to try and catch here ?

Reply:
Anything that throws in the function will get caught on line 121 and the error will be printed to the console. Are you thinkoing of any other error handling. For me that seemed enough, as there is nothing else custom that I'd want to do.

Reply:
Anything that throws in the function will get caught on line 121 and the error will be printed to the console. Are you thinkoing of any other error handling. For me that seemed enough, as there is nothing else custom that I'd want to do.